### PR TITLE
Zoltan2:Zoltan2xtrapulp - integration of XtraPuLP-0.3 into Zoltan2 for multi-weight partitioning

### PR DESCRIPTION
**Pull request by gslota**
@trilinos/zoltan2 

## Motivation
Integration of multi-weight partitioning for XtraPuLP-0.3. 
Updates to Zoltan2_AlgPuLP.hpp and relevant testing files.

## Testing
'ctest' completes successfully.

## Additional Information
Most recent XtraPuLP version available via: https://github.com/HPCGraphAnalysis/PuLP

### CODE REVIEW

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp [hunk 1]
```diff
@@ -342,11 +351,20 @@ void AlgPuLP<Adapter>::partition(
...
+    scale_weights(nVtx, nVwgts, vwgts, vertex_weights);
     for (int i = 0; i < num_verts; ++i)
       vertex_weights_sum += vertex_weights[i];
   }
+#else
+  // XtraPuLP supports an arbitrary number of vertex weights
+  int* vertex_weights = NULL;
+  if (nVwgts) {
+    vertex_weights = new int[nVtx*nVwgts];
+    scale_weights(nVtx, nVwgts, vwgts, vertex_weights);
```
Comment (mhoemmen) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp]: Does this leak memory for `vertex_weights`?  See note above.

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp [hunk 2]
```diff
@@ -252,8 +256,9 @@ class AlgPuLP : public Algorithm<Adapter>
 
   void buildModel(modelFlag_t &flags);
 
-  void scale_weights(size_t n, StridedData<lno_t, scalar_t> &fwgts,
-                     int *iwgts);
+  void scale_weights( size_t n, int nWgt,
```
Comment (mhoemmen) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp]: 1. This interface does not make clear whether it transfers ownership of `iwgts`.  That lack of clarity could cause memory leaks or double frees.  If you intend to transfer ownership to the callee, then you should pass `iwgts` as `unique_ptr<int[]>`.

2. Why are you passing `ArrayView` by nonconst reference?  That suggests that it's an output parameter.  If so, then why not just return `ArrayView`?

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp [hunk 3]
```diff
@@ -342,11 +351,20 @@ void AlgPuLP<Adapter>::partition(
...
-    scale_weights(nVtx, vwgts[0], vertex_weights);
+    scale_weights(nVtx, nVwgts, vwgts, vertex_weights);
     for (int i = 0; i < num_verts; ++i)
       vertex_weights_sum += vertex_weights[i];
   }
+#else
+  // XtraPuLP supports an arbitrary number of vertex weights
+  int* vertex_weights = NULL;
+  if (nVwgts) {
+    vertex_weights = new int[nVtx*nVwgts];
```
Comment (mhoemmen) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp]: Use of raw `new` and `delete` is an antipattern.  Prefer `unique_ptr` if you mean to transfer ownership to the callee (`scale_weights`).

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp [hunk 4]
```diff
@@ -492,33 +535,50 @@ void AlgPuLP<Adapter>::partition(
...
+    do_edge_bal, do_maxcut_min,
+    verbose_output, pulp_seed};
+
   ierr = pulp_run(&g, &ppc, parts, num_parts);  
 
   env->globalInputAssertion(__FILE__, __LINE__, "pulp_run", 
     !ierr, BASIC_ASSERTION, problemComm);
 #else
+  // Create XtraPuLP's partitioning data structure
+  double* constraints = NULL;
```
Comment (mhoemmen) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp]: Prefer `nullptr` to `NULL`.  Also, use of raw `new` and `delete` is an antipattern.

File: packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp (hunk 2 above)
Comment (gslota) [reply packages/zoltan2/src/algorithms/partition/Zoltan2_AlgPuLP.hpp]: I've addressed this in a recent commit. The function was taken as-is from other partitioning methods, but I've updated it within Zoltan2_AlgPuLP to make the interface more clear.

**Comment by cgcgcg** [comment]
@gslota I think you want to rebase your branch..

**Comment by lucbv** [comment]
@gslota thanks for rebasing yesterday, this looks good now!

**Comment by kddevin** [comment]
@cgcgcg Thanks for taking a second look.  If you are satisfied, would you please change your review to remove the block?  Thanks.

**Pull request merged**